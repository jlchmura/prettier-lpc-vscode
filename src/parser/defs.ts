export const tt = {
  _HSH: "#".charCodeAt(0),
  _BNG: "!".charCodeAt(0),
  _DLR: "$".charCodeAt(0),
  _MIN: "-".charCodeAt(0),
  _LAN: "<".charCodeAt(0),
  _RAN: ">".charCodeAt(0),
  _BSL: "\\".charCodeAt(0),
  _FSL: "/".charCodeAt(0),
  _EQS: "=".charCodeAt(0),
  _DQO: '"'.charCodeAt(0),
  _SQO: "'".charCodeAt(0),
  _NWL: "\n".charCodeAt(0),
  _CAR: "\r".charCodeAt(0),
  _LFD: "\f".charCodeAt(0),
  _WSP: " ".charCodeAt(0),
  _TAB: "\t".charCodeAt(0),
  _SEM: ";".charCodeAt(0),
  _STR: "*".charCodeAt(0),
  _OPP: "(".charCodeAt(0),
  _CLP: ")".charCodeAt(0),
  _PLS: "+".charCodeAt(0),
  _MNS: "-".charCodeAt(0),
  _OBK: "{".charCodeAt(0),
  _CBK: "}".charCodeAt(0),
  _OSB: "[".charCodeAt(0),
  _CSB: "]".charCodeAt(0),
  _DOT: ".".charCodeAt(0),
  _COM: ",".charCodeAt(0),
  _COL: ":".charCodeAt(0),
  _PIP: "|".charCodeAt(0),
  _AMP: "&".charCodeAt(0),
  _QUE: "?".charCodeAt(0),
  _PCT: "%".charCodeAt(0),
  _AT: "@".charCodeAt(0),
  _IN: "in ".split("").map((c) => c.charCodeAt(0)),  // must have a space after it
  _CALLOTHER: "->".split("").map((c) => c.charCodeAt(0)),
  _INHERIT: "inherit".split("").map((c) => c.charCodeAt(0)),
  _IF: "if".split("").map((c) => c.charCodeAt(0)),
  _ELSEIF: "else if".split("").map((c) => c.charCodeAt(0)),
  _ELSE: "else".split("").map((c) => c.charCodeAt(0)),
  _RETURN: "return".split("").map((c) => c.charCodeAt(0)),
  _FOR: "for".split("").map((c) => c.charCodeAt(0)),
  _FOREACH: "foreach".split("").map((c) => c.charCodeAt(0)),
  _WHILE: "while".split("").map((c) => c.charCodeAt(0)),
  _BREAK: "break".split("").map((c) => c.charCodeAt(0)),
  _CONTINUE: "continue".split("").map((c) => c.charCodeAt(0)),
  _SWITCH: "switch".split("").map((c) => c.charCodeAt(0)),
  _CASE: "case".split("").map((c) => c.charCodeAt(0)),
  _DEFAULT: "default".split("").map((c) => c.charCodeAt(0)),
  _LAMBDA: "lambda".split("").map((c) => c.charCodeAt(0)),
  _SPREAD: "...".split("").map((c) => c.charCodeAt(0)),
};

export const binary_ops_list = [
  "==",
  "!=",
  "<<<",
  "<=",
  "<",
  ">=",
  ">",
  "<<",
  ">>",
];
export const binary_ops = binary_ops_list.map((o) => o.split("").map((c) => c.charCodeAt(0)));
export const binary_ops_set = new Set(binary_ops_list);

export const logical_ops_set = new Set(["||", "&&"]);

export const logical_ops = ["||", "&&"].map((o) =>
  o.split("").map((c) => c.charCodeAt(0))
);

export const assignment_ops = [
  "=",
  "+=",
  "-=",
  "&=",
  "|=",
  "^=",
  "<<=",
  ">>=",
  ">>>=",
  "*=",
  "%=",
  "/=",
  "&&=",
  "||=",
  "++",
  "--"  
].map((o) => o.split("").map((c) => c.charCodeAt(0)));

export const arrith_ops = ["+", "-", "*", "/", "!", "%", "&", "|"].map((o) =>
  o.split("").map((c) => c.charCodeAt(0))
);

export const unary_ops_set = new Set(["!", "-", "&", "++", "--"]);

export enum DeclType {
  Function = 1,
  Variable = 2,
  Inherit = 4,
}

export const op_precedence: { [op: string]: number } = {
  "||": 1,
  "&&": 2,
  "|": 3,
  "^": 4,
  "&": 5,
  "==": 6,
  "!=": 6,
  "<": 7,
  "<=": 7,
  ">": 7,
  ">=": 7,
  "<<": 8,
  ">>": 8,
  "+": 9,
  "-": 9,
  "%": 10,
  "*": 10,
  "/": 10,
};

export const modifiers: { [mod: string]: number } = {
  private: DeclType.Function | DeclType.Variable,
  protected: DeclType.Function,
  static: DeclType.Function | DeclType.Variable,
  public: DeclType.Function | DeclType.Variable,
  nomask: DeclType.Function,
  varargs: DeclType.Function,
  deprecated: DeclType.Function | DeclType.Variable,
  virtual: DeclType.Inherit,
  nosave: DeclType.Variable
};

export const types = [
  "int",
  "string",
  "status",
  "object",
  "array",
  "mapping",
  "closure",
  "symbol",
  "float",
  "mixed",
  "struct",
  "union",
  "null",
  "void",
  "class"
];
export const typesSet = new Set(types);

export const typesAllowAsVarNames = [
  "status",
  "symbol",
  "null"
];
export const typesAllowAsVarNamesSet = new Set(typesAllowAsVarNames);

export const modifiers_function = [
  "private",
  "protected",
  "static",
  "public",
  "nomask",
  "varargs",
  "deprecated",
];
export const modifiers_variable = [
  "private",
  "nosave",
  "static",
  "public",
  "deprecated",
];
export const modifiers_inherit = ["virtual"];

export const modifiers_set = new Set(
  modifiers_function.concat(modifiers_variable).concat(modifiers_inherit)
);

export const directives = [
  "include",
  "define",
  "ifdef",
  "ifndef",
  // must come after other if* directives
  "if",
  "else",
  "elif",
  "endif",
  "undef",
  "line",
  "echo",
  "pragma",
];
export const directive_chars = directives.map((o) =>
  o.split("").map((c) => c.charCodeAt(0))
);

export const reserved_words = [
  "switch",
  "case",
  "break",
  "continue",
  "struct",
  "union",
  "enum",
  "default",
  "int",
  "char",
  "float",
  "double",
  "typdef",
  "unsigned",
  "register",
  "static",
  "global",
  "extern",
  "void",
  "inherit",
  //"va_dcl",
  "goto",
  "return",
  "if",
  "while",
  "for",
  "foreach",
  "else",
  "do",
  //"sizeof",
];
export const reserved_words_set = new Set(reserved_words);
