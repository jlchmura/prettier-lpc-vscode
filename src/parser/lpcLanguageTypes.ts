export enum TokenType {
  None = 0,

  StartCommentBlock,
  EndCommentBlock,
  Comment,

  Semicolon,
  Colon,
  Comma,

  Directive,
  DirectiveArgument,
  DirectiveLineBreak,
  DirectiveEnd,

  Inherit,
  InheritanceAccessor,

  StartDeclaration,
  Modifier,
  Type,
  DeclarationName,

  TypeCast,

  CodeBlockStart,
  CodeBlockEnd,

  ArrayStart,
  ArrayEnd,

  MappingStart,
  MappingEnd,

  Variable,

  Arrow,

  Function,
  FunctionArgument,
  FunctionArgumentType,
  FunctionArgumentEnd,
  Return,

  Closure,
  InlineClosureStart,
  InlineClosureEnd,
  InlineClosureArgument,
  LambdaStart,
  LabmdaEnd,
  LambdaEmptyArg,
  LambdaIndexor,  

  If,
  ElseIf,
  Else,
  Expression,
  Ternary,

  Switch,
  SwitchCase,
  
  For,
  ForEach,
  ForEachIn,
  ForEachRange,
  While,
  ControlFlow,

  Literal,
  LiteralChar,
  LiteralNumber,

  ParenBlock,
  ParenBlockEnd,

  IndexorStart,
  IndexorEnd,
  IndexorFromEndPos,
  IndexorPosSep,

  StructLiteral,

  Operator,
  LogicalOperator,
  AssignmentOperator,
  Bang,
  Star,
  CallByRef,
  Spread, // FluffOS only

  Whitespace,
  BlankLines,

  StringLiteralStart,
  StringLiteralBody,

  Unknown,
  EOS,
}

export enum ScannerState {
  WithinFile,

  StartDirective,
  WithinDirective,
  WithinDirectiveArg,

  StartInherit,
  WithinInherit,

  StartDeclaration,
  WithinFunction,
  WithinFunctionArgs,
  WithinAssignment,

  WithinLiteral,

  WithinBlock,

  WithinCommentBlock,

  WithinLambda,

  IfExpression,
  ElseIfExpression,
  ElseExpression,
  WithinIf,

  WithinIndexor,

  WithinStringLiteralBlock
}
