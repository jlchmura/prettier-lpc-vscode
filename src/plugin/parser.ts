import {
  AstPath,
  Doc,
  Parser,
  ParserOptions,
  Printer,
  RequiredOptions,
} from "prettier";
import { builders } from "prettier/doc";
import { LPCOptions } from ".";
import { LPCNode } from "../nodeTypes/lpcNode";
import { LPCParser } from "../parser/lpcParser";


export const locStart = (node: LPCNode) => node.start;
    export const locEnd =  (node:LPCNode) => node.end;
    export const parse = (
      text: string,
      parsers: { [parserName: string]: Parser },
      options: LPCOptions
    ): LPCNode => {
      const parse = new LPCParser();
      const doc = parse.parse(text);
      const nd = new LPCNode(0, text.length-1, doc.roots, void 0) ;
      nd.type = "root";
      return nd;
    }