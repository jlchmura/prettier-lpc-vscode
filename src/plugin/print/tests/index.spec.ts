import * as prettierPlugin from "../..";
import * as prettier from "prettier";
import { LPCParser, ParseLPC } from "../../../parser/lpcParser";
import {
  assign_exp_suffix_comment,
  for_loop_various,
  if_condense_test,
  literal_consecutive_strings,
  mapping_with_ternary_value,
  spec_input_room,
} from "./inputs";

describe("prettier-lpc plugin", () => {
  const format = (
    input: string,
    options?: Partial<prettierPlugin.LPCOptions>
  ) => {
    const formatted = prettier.format(input, {
      parser: prettierPlugin.AST_PARSER_NAME,
      plugins: [prettierPlugin],
      printWidth: 80,
      tabWidth: 2,
      useTabs: false,
      ...options,
    });

    // always validating that the result of formatting is backward compatible with JSONata
    expect(() => ParseLPC(formatted)).not.toThrow();
    return formatted;
  };

  test("formats short functions", () => {
    let formatted = format(`string set_s(string s){return foo=s;}`);
    expect(formatted).toMatchInlineSnapshot(
      `"string set_s(string s) { return foo = s; }"`
    );

    formatted = format(`query_s ( ) {return s;}`);
    expect(formatted).toMatchInlineSnapshot(`"query_s() { return s; }"`);

    formatted = format(`test() { int i=0; return 1; }`);
    expect(formatted).toMatchInlineSnapshot(`
      "test() {
        int i = 0;
        return 1;
      }"
    `);
  });

  test("format array typecasts", () => {
    let formatted = format(`string *dirs = (string *) env->query_dest_dir();`);
    expect(formatted).toMatchInlineSnapshot(
      `"string *dirs = (string*)env->query_dest_dir();"`
    );

    formatted = format(
      `string * dirs = obj->fn( (string *) env->query_dest_dir() );`
    );
    expect(formatted).toMatchInlineSnapshot(
      `"string *dirs = obj->fn((string*)env->query_dest_dir());"`
    );
  });

  test("format arrays", () => {
    let formatted = format(
      `test() { items = ({     ({"gates", "steel gates"}), "Strong metal gates.", }); }`
    );
    expect(formatted).toMatchSnapshot("array-condensed");

    formatted = format(
      `test() { items = ({     
        ({"gates", "steel gates"}), "Strong metal gates.", }); }`
    );
    expect(formatted).toMatchSnapshot("array-inner-condensed");

    formatted = format(
      `test() { items = ({     
        ({ 
          "gates", "steel gates"}), "Strong metal gates.", }); }`
    );
    expect(formatted).toMatchSnapshot("array-none-condensed");
  });

  test("format args passed byref", () => {
    let formatted = format(`test(string &d) { d[0] += 32; }`);
    expect(formatted).toMatchInlineSnapshot(
      `"test(string &d) { d[0] += 32; }"`
    );
  });

  test("format ternary expressions", () => {
    // without paren
    let formatted = format(
      `printf("Foo is %s\\n", test=="bar" ? "bar" : "notbar");`
    );
    expect(formatted).toMatchInlineSnapshot(
      `"printf("Foo is %s\\n", test == "bar" ? "bar" : "notbar");"`
    );

    // with paren
    formatted = format(
      `printf("Foo is %s\\n", (test=="bar") ? "bar" : "notbar");`
    );
    expect(formatted).toMatchInlineSnapshot(
      `"printf("Foo is %s\\n", (test == "bar") ? "bar" : "notbar");"`
    );

    // arith op inside ternary
    formatted = format(
      `private int round(float n) { return (int)(n < 0 ? n - 0.5 : n + 0.5); }`
    );
    expect(formatted).toMatchInlineSnapshot(
      `"private int round(float n) { return (int)(n < 0 ? n - 0.5 : n + 0.5); }"`
    );

    formatted = format(
      `private int round(float n) { int i=(int)(n < 0 ? n - 0.5 : n + 0.5); return i; }`
    );
    expect(formatted).toMatchSnapshot("ternary_arith_op_inside");

    formatted = format(
      `test ()
{
  if (!a && !ob->b())
    return "foo";
  else
    return sprintf("bar %s",
          ob->test() && ob2->test() 
          ? "bar"
        : ob->bar() ? "baz" : "baz2");
}
`
    );
    expect(formatted).toMatchInlineSnapshot(`
          "test() {
            if (!a && !ob->b()) return "foo";
            else
              return
                sprintf(
                  "bar %s",
                  ob->test() && ob2->test() ? "bar" : ob->bar() ? "baz" : "baz2"
                );
          }"
        `);

    formatted = format(mapping_with_ternary_value);
    expect(formatted).toMatchSnapshot("mapping_with_ternary_value");

    formatted = format(
      `test() { str = (str[0] == '/' ? "/" + implode(path, "/") : implode(path, "/")) }`
    );
    expect(formatted).toMatchSnapshot("ternary-after-binary");
  });

  test("format binary expressions", () => {
    let formatted = format(
      `test() { if ( a != 0 && ( b == 0 || c >= MIN )) {return 1;}}`
    );
    expect(formatted).toMatchInlineSnapshot(`
      "test() {
        if (a != 0 && (b == 0 || c >= MIN)) {
          return 1;
        }
      }"
    `);
  });

  test("format logical expression", () => {
    let formatted = format(`test(str) {return str == "NO" || str == "TWO";}`);
    expect(formatted).toMatchInlineSnapshot(
      `"test(str) { return str == "NO" || str == "TWO"; }"`
    );

    formatted = format(
      `test() { if (str == "a" && found2 & !taken) { write("You find nothing."); return 1; } }`
    );
    expect(formatted).toMatchInlineSnapshot(`
      "test() {
        if (str == "a" && found2 & !taken) {
          write("You find nothing.");
          return 1;
        }
      }"
    `);
  });

  test("format labmdas", () => {
    let formatted = format(`test() { 
      list = filter_array(info(), lambda(({ 'test }), ({ #'!=, ({ #'[, 'isValid, NAME }), \"NONAME\" }) )); 
    }`);
    expect(formatted).toMatchInlineSnapshot(`
          "test() {
            list =
              filter_array(
                info(),
                lambda(({'test}), ({#'!=, ({#'[, 'isValid, NAME}), "NONAME"}))
              );
          }"
        `);
  });

  test("format closures", () => {
    let formatted = format(
      `test() { object *hash = ({}); hash = sort_array(hash, #'>); hash = filter(hash, #'this_player); }`
    );
    expect(formatted).toMatchSnapshot("closure-greaterthan-this_player");

    formatted = format(
      `object *a = filter(all_inventory(room), (: $1->id("something") :));`
    );
    expect(formatted).toMatchInlineSnapshot(
      `"object *a = filter(all_inventory(room), (: $1->id("something") :));"`
    );

    formatted = format(`int *arr=filter(arr2,(:($1==1&&$1<10):));`);
    expect(formatted).toMatchInlineSnapshot(
      `"int *arr = filter(arr2, (: ($1 == 1 && $1 < 10) :));"`
    );
  });

  test("format variable declarations", () => {
    let formatted = format(`string 
      *arr,		/* OPTIONAL: Array of stuff */
      code;			/* test comment */
    `);

    expect(formatted).toMatchInlineSnapshot(`
          "string *arr, /* OPTIONAL: Array of stuff */
                 code; /* test comment */"
        `);

    // multi-decl var
    formatted = format(`string *arr, code="", test=0, s;`);
    expect(formatted).toMatchInlineSnapshot(
      `"string *arr, code = "", test = 0, s;"`
    );

    // multilpe types of var decls
    formatted = format(`string s; mixed m; int i=0; mapping mp;`);
    expect(formatted).toMatchInlineSnapshot(`
      "string s;
      mixed m;
      int i = 0;
      mapping mp;"
    `);

    // var decl inside function
    formatted = format(`test() { int i, j; string s=""; }`);
    expect(formatted).toMatchInlineSnapshot(`
      "test() {
        int i, j;
        string s = "";
      }"
    `);
  });

  test("format struct definitions", () => {
    let formatted = format(`// this struct should be on a single line
    struct coords { int x; int y; };`);

    expect(formatted).toMatchInlineSnapshot(`
          "// this struct should be on a single line
          struct coords { int x; int y; };"
        `);
  });

  test("format macros", () => {
    let formatted = format(
      `#define WRAP(str)  trim("test"->word_wrap(str), 2)`
    );
    expect(formatted).toMatchInlineSnapshot(
      `"#define WRAP(str) trim("test"->word_wrap(str), 2)"`
    );

    formatted = format(
      `#define WRAP(str)  trim("really really really really long string"->word_wrap(str), 2)`,
      { printWidth: 20 }
    );
    expect(formatted).toMatchSnapshot("define-macro-with-wrap");

    formatted = format(`#define W2(s) \
      trim("local/really really really long string really really really really long/util")-> \
      wrap(({ a, s }))
    `);
    expect(formatted).toMatchSnapshot("define-macro-multiline");
  });

  test("format arrow operators", () => {
    let formatted = format(`test() {
      "/obj/master"->
        query_player_exists();
    }`);

    expect(formatted).toMatchSnapshot("arrow-newline-after");

    // arrow after obj
    formatted = format(`string s = obj->test();`);
    expect(formatted).toMatchInlineSnapshot(`"string s = obj->test();"`);

    // arrow after call-exp
    formatted = format(`string s=fn()->test();`);
    expect(formatted).toMatchInlineSnapshot(`"string s = fn()->test();"`);

    // arrow after binary-exp in parens
    formatted = format(`string s=("obj"+"name")->test();`);
    expect(formatted).toMatchInlineSnapshot(
      `"string s = ("obj" + "name")->test();"`
    );

    // arrow inside ternary after arrow
    formatted = format(`string s = obj->test()?obj->test2() : "";`);
    expect(formatted).toMatchInlineSnapshot(
      `"string s = obj->test() ? obj->test2() : "";"`
    );

    // arrow with newlines
    formatted = format(`string s = obj
    ->
    test();`);
    expect(formatted).toMatchInlineSnapshot(`"string s = obj->test();"`);

    // arrow with suffix comment
    formatted = format(`string s = obj->fn(); // comment`);
    expect(formatted).toMatchInlineSnapshot(
      `"string s = obj->fn(); // comment"`
    );
  });

  test("format foreach loops", () => {
    let formatted = format(
      `test() { string exitKey; foreach(exitKey : all_exits) { write(exitKey); } }`
    );
    expect(formatted).toMatchSnapshot("foreach-collapsed");

    formatted = format(
      `test() { string exitKey; foreach(exitKey : all_exits) { write(exitKey); i++; } }`
    );
    expect(formatted).toMatchSnapshot(`foreach-multiline`);

    formatted = format(`test() { foreach(i : 1 .. 6) printf("%d\n", i); }`);
    expect(formatted).toMatchSnapshot("foreach-range-collapsed");

    formatted = format(
      `test() { foreach(i : 1 .. 6) { printf("%d\n", i); j++; } }`
    );
    expect(formatted).toMatchSnapshot("foreach-range-multiline");
  });

  test("general formatting", () => {
    let formatted = format(spec_input_room);
    expect(formatted).toMatchSnapshot("spec_input_room");
  });

  test("format functions", () => {
    let formatted = format(
      `int level=0; public int query_level(); public void set_level(int level); public int query_next_level(); public int query_level() { return level ; }`
    );
    expect(formatted).toMatchSnapshot("function-stubs");

    formatted = format(`int gmcp_send_map();
    int gmcp_send_map_config(); 
    
    /**
     * Should not move the coment up after the semi
     */`);
    expect(formatted).toMatchSnapshot("function-stub-with-newline-comments");
  });

  test("format assignment expressions", () => {
    let formatted = format(assign_exp_suffix_comment);
    expect(formatted).toMatchSnapshot("assign_exp_suffix_comment");
  });

  test("formats if statements", () => {
    let formatted = format(if_condense_test);
    expect(formatted).toMatchSnapshot("if_condense_test");
  });

  test("formats for loops", () => {
    let formatted = format(
      `test() { for (i = 0, j = sizeof(keys); i < j; i++, j--) { string key = keys[i];     }}`
    );
    expect(formatted).toMatchSnapshot("for-loop_multi_expression");

    formatted = for_loop_various;
    expect(formatted).toMatchSnapshot("for_loop_various");
  });

  test("formatter should handle missing semi's", () => {
    // comma instead of semi
    let formatted = format(
      `test() { short = "short name", long = "long" + "desc"; }`
    );
    expect(formatted).toMatchSnapshot("missing_semi_comma_instead");
  });

  test("format literal strings", () => {
    // consecutive strings with no +
    let formatted = format(literal_consecutive_strings);
    expect(formatted).toMatchSnapshot("literal_consecutive_strings");
  });
});
