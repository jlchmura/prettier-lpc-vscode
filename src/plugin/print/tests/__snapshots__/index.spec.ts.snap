// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettier-lpc plugin fluffos text formatting shortcuts (@): textFormattingDouble 1`] = `
"void test() {
  set_desc(@@TXT
This is
  a tes
TXT
  );
}"
`;

exports[`prettier-lpc plugin fluffos text formatting shortcuts (@): textFormattingSingle 1`] = `
"void test() {
  set_desc(@TXT
This is
  a test
TXT
  );
}"
`;

exports[`prettier-lpc plugin format arrays: array-condensed 1`] = `"test() { items = ({({"gates", "steel gates"}), "Strong metal gates."}); }"`;

exports[`prettier-lpc plugin format arrays: array-inner-condensed 1`] = `
"test() {
  items = ({
    ({"gates", "steel gates"}),
    "Strong metal gates.",
  });
}"
`;

exports[`prettier-lpc plugin format arrays: array-none-condensed 1`] = `
"test() {
  items = ({
    ({
      "gates",
      "steel gates",
    }),
    "Strong metal gates.",
  });
}"
`;

exports[`prettier-lpc plugin format arrays: array-with-multi-sfx-comments 1`] = `
"
test() {
  items = ({
    ({"gates", "steel gates"}),
    "Strong metal gates.", // inner sfx
  }); // sfx
  // post
}"
`;

exports[`prettier-lpc plugin format arrays: array-with-sfx-comment 1`] = `
"
test() {
  items = ({({"gates", "steel gates"}), "Strong metal gates."}); // sfx
}"
`;

exports[`prettier-lpc plugin format arrow operators: arrow-newline-after 1`] = `
"test() {
  "/obj/master"->query_player_exists();
}"
`;

exports[`prettier-lpc plugin format assignment expressions: assign_exp_suffix_comment 1`] = `
"test() {
  exits = room->exits();
  if (exits == -1) arr = ([]); // erase all exits
  else if (stringp(exits)) arr -= ([exits]);
}"
`;

exports[`prettier-lpc plugin format assignment expressions: assignment_inside_array 1`] = `
"int i = 0;
test() { indices = ({i, index++}); }"
`;

exports[`prettier-lpc plugin format binary expressions: call-exp-inside-binary-string-exp 1`] = `
"string test() {
  string name = query_name();
  return name + " was defeated by " + "/daemons/time_d"->query_time() + ".
";
}"
`;

exports[`prettier-lpc plugin format call-exp inside arrays: call-exp-in-array 1`] = `
"#define SOME_DEFINE "some define"

void somefunc() {
  string *words;
  words = ({SOME_DEFINE, lower_case(SOME_DEFINE), upper_case(SOME_DEFINE)});
}"
`;

exports[`prettier-lpc plugin format closures: closure-greaterthan-this_player 1`] = `
"test() {
  object *hash = ({});
  hash = sort_array(hash, #'>);
  hash = filter(hash, #'this_player);
}"
`;

exports[`prettier-lpc plugin format foreach loops: foreach-collapsed 1`] = `
"test() {
  string exitKey;
  foreach (exitKey : all_exits) { write(exitKey); }
}"
`;

exports[`prettier-lpc plugin format foreach loops: foreach-multi-var 1`] = `"test() { foreach (x, y : a) { z = b[x]; } }"`;

exports[`prettier-lpc plugin format foreach loops: foreach-multi-var-keep-in 1`] = `"test() { foreach (x, y in a) { z = b[x]; } }"`;

exports[`prettier-lpc plugin format foreach loops: foreach-multiline 1`] = `
"test() {
  string exitKey;
  foreach (exitKey : all_exits) {
    write(exitKey);
    i++;
  }
}"
`;

exports[`prettier-lpc plugin format foreach loops: foreach-range-collapsed 1`] = `
"test() { foreach (i : 1 .. 6) printf("%d
", i); }"
`;

exports[`prettier-lpc plugin format foreach loops: foreach-range-multiline 1`] = `
"test() {
  foreach (i : 1 .. 6) {
    printf("%d
", i);
    j++;
  }
}"
`;

exports[`prettier-lpc plugin format functions: function-stub-with-newline-comments 1`] = `
"int gmcp_send_map();
int gmcp_send_map_config();

/* Should not move the coment up after the semi */"
`;

exports[`prettier-lpc plugin format functions: function-stubs 1`] = `
"int level = 0;
public int query_level();
public void set_level(int level);
public int query_next_level();
public int query_level() { return level; }"
`;

exports[`prettier-lpc plugin format literal strings: literal_consecutive_strings 1`] = `
"test() {
  desc =
    "this is line one. it is the first line in the description
" +
    "this is line 2.
" +
    "this is line 3.";

  desc =
    "this is line one. it is the first line in the description
" +
    "this is line 2.
" +
    "this is line 3.";
}"
`;

exports[`prettier-lpc plugin format macros: define-macro-multiline 1`] = `
"#define W2(s) \\
  trim("local/really really really long string really really really really long/util")->       wrap(({ a, s }))"
`;

exports[`prettier-lpc plugin format macros: define-macro-with-wrap 1`] = `
"#define WRAP(str) \\
  trim("really really really really long string"->word_wrap(str), 2)"
`;

exports[`prettier-lpc plugin format ternary expressions: mapping_with_ternary_value 1`] = `
"/**
* send data */
public varargs int send_config(object player) {
  mixed *arr = ({}); // array of environment data objects

  mapping data = MASTER->query_data();
  if (!data) {
    write("WARN: Could not load");
  }
  foreach (string nm : data) {
    int id = data[nm, 0];
    mapping d = ([
      ID:
        id ? id : "0", // need string zero, not int 0
      NAME: nm,
      C: data[nm, 1],
    ]);

    arr += ({d});
  }

  // package
  mapping pkg = ([KEY: arr]);

  object p = player ? player : this_player();
  return p->send(PKG, pkg);
}"
`;

exports[`prettier-lpc plugin format ternary expressions: ternary_arith_op_inside 1`] = `
"private int round(float n) {
  int i = (int)(n < 0 ? n - 0.5 : n + 0.5);
  return i;
}"
`;

exports[`prettier-lpc plugin format ternary expressions: ternary-after-binary 1`] = `
"test() {
  str = (str[0] == '/' ? "/" + implode(path, "/") : implode(path, "/"));
}"
`;

exports[`prettier-lpc plugin format ternary expressions: ternary-with-lit-binary-op 1`] = `
"test() {
  set_long(
    "The " + race_arr[race] +
    " looks at " + (gender == 1 ? "him" : "her") +
    "."
  );
}"
`;

exports[`prettier-lpc plugin formats for loops: for_loop_various 1`] = `
"test() {
  for (int i=0,j=1;i<10;i++,j+=2) { fn(i,j); }
  for (int i=0,j=1;i<10;) { fn(i,j); }
  int i;
  for (i=0;i<10;i--) { fn(i); }
}"
`;

exports[`prettier-lpc plugin formats for loops: for-loop_multi_expression 1`] = `
"test() {
  for (i = 0, j = sizeof(keys); i < j; i++, j--) { string key = keys[i]; }
}"
`;

exports[`prettier-lpc plugin formats for loops: loop-break-shouldhave-semi 1`] = `"void test() { while (true) { if (1 == 1) break; } }"`;

exports[`prettier-lpc plugin formats for loops: loop-continue-shouldhave-semi 1`] = `"void test() { while (true) { if (1 == 1) continue; } }"`;

exports[`prettier-lpc plugin formats if statements: if_condense_test 1`] = `
"test() {
  // this should be fairly condensed
  if (
    str != "down" && str != "up" && str != "hole" && str != "hill" &&
    str != "mountain"
  ) return 0;
  else if (str == "sky") {
    // send them
    move_player("sky");
    return 1;
  } else return 0;
}"
`;

exports[`prettier-lpc plugin formatter should handle missing semi's: missing_semi_comma_instead 1`] = `
"test() {
  short = "short name";
  long = "long" + "desc";
}"
`;

exports[`prettier-lpc plugin general formatting: spec_input_room 1`] = `
"#include "../globals.h"
inherit "/room/room";

object monster;

reset(arg) {
  if (!arg) {
    set_light(1);
    set_realm("something");
    short_desc = "Room";
    long_desc =
      "This is a long
" +
      "Room description that goes
" +
      "on for several lines
";
    items = ({
      ({"table", "small table", "wood table"}),
      "A table made of wood",
      ({"bench", "wood bench"}),
      "A bench made of wood",
    });

    dest_dir = ({
      BASE + "/northroom",
      "north",
    });
  }

  if (!present("monster")) {
    monster = clone_object("obj/monster");
    monster->set_name("monster");
    monster->set_alias(({"monster"}));
    monster->set_hp(200);
    move_object(monster, this_object());
  }
}"
`;

exports[`prettier-lpc plugin handle string literals with escaped quotes: literal_strings_multiple_escapes 1`] = `"string s = "testing\\" 123\\\\";"`;

exports[`prettier-lpc plugin handles the spread operator (FluffOS): spread-op-function-and-callexp 1`] = `
"mixed sum(mixed *numbers...) {
  mixed number, result = 0;
  fn(1, numbers...);
  return result;
}"
`;
